### DISCUSSION TITLE
Display VSTi parameters from Cubase

### DISCUSSION THREAD

**Discussion by spamanski** [tttapa/Control-Surface#1079] [Q&A] [answered]
Hi there, 

first of all, thanks a lot for your great work, Pieter! 
I've built quite a few basic controllers with your library and they are all working great. 

Now I would like to take it a step further and have some visual feedback on the controller. 

Basically, what I want to achieve is this: 
![mcu_cubase_instr_ed](https://github.com/user-attachments/assets/056ff829-810a-48db-8982-b2a4dee4c29f)

It's very easy to control these parameters with a generic remote, not so easy with emulating a real MCU unit, but one or another, the hardest thing is the display integration (via MCU, not a generic controller of course). 

As far as I understand, it's possible to display pretty much anything being sent from the controller to the DAW. And there is also a bunch of incoming stuff defined, like MCU::LCD or MCU::TimeDisplay, although it seems like there is no complete list of things the controller can "listen to" and draw on the display (or is there?). 
So I really struggle to understand how to get the parameter values (and possibly names) from Cubase to the displays on the controller. 

I tried the reverse engineering sketch and also MidiOx with loopback ports, but I can't see anything sent out of Cubase to the controller when opening or tweaking these controls. I can see regular note and CC events for the usual stuff though, also sysex when opening and closing the DAW. 

I know you're not a Cubase user, but do you have any idea how to make this work? I have studied your files and documentations for days, but this goes over my head. 

Also, am I right to assume that the displays only show what the controller tells them to? 
Like... (sorry, I don't know the terminology) - it's not like the display dynamically shows whatever is in "focus" on the DAW because of some kind of feedback the display receives from the DAW?

Thanks so much in advance :)

**Comment by tttapa** [comment, marked as answer]
> I can see regular note and CC events for the usual stuff though, also sysex when opening and closing the DAW.

The display contents are sent using SysEx, so it's best to have a closer look at those messages. The format is explained here: https://tttapa.github.io/Control-Surface-doc/Doxygen/d5/d2e/classMCU_1_1LCD.html

> Also, am I right to assume that the displays only show what the controller tells them to?

Indeed, the DAW just sends a SysEx message to the controller that says something along the lines of: "please write the characters 'XYZ' to position _n_ on the LCD display". The controller itself does not do any text processing, it just displays the characters it received. These characters can be accessed using the `MCU::LCD::getText()` function.

Here's an example, integrated with the rest of the reverse engineering code:

```cpp
#include <Control_Surface.h>

// A MIDI over USB interface
USBMIDI_Interface midi;

bool channelMessageCallback(ChannelMessage cm) {
  MIDIMessageType type = cm.getMessageType();
  if (type == MIDIMessageType::NoteOn || type == MIDIMessageType::NoteOff) {
    Serial << hex << cm.header << ' ' << cm.data1 << ' ' << cm.data2 << dec
           << "\t(" << MCU::getMCUNameFromNoteNumber(cm.data1) << ")"
           << F(" on cable ") << cm.cable.getOneBased() << endl;
  } else {
    Serial << hex << cm.header << ' ' << cm.data1 << ' ' << cm.data2 << dec
           << F(" on cable ") << cm.cable.getOneBased() << endl;
  }
  return false;
}

bool sysExMessageCallback(SysExMessage se) {
  Serial << F("System Exclusive message: [") << se.length << "] " //
         << AH::HexDump(se.data, se.length)                       //
         << F(" on cable ") << se.cable.getOneBased() << endl;
  return false;
}

bool sysCommonMessageCallback(SysCommonMessage sc) {
  Serial << F("System Common message: ") << hex << sc.getMessageType();
  if (sc.getNumberOfDataBytes() >= 1)
    Serial << sc.getData1();
  if (sc.getNumberOfDataBytes() >= 2)
    Serial << sc.getData2();
  Serial << dec << F(" on cable ") << sc.cable.getOneBased() << endl;
  return false;
}

bool realTimeMessageCallback(RealTimeMessage rt) {
  Serial << F("Real-Time: ") << hex << rt.message << dec << F(" on cable ")
         << rt.cable.getOneBased() << endl;
  return false;
}

MCU::LCD<> lcd;

void setup() {
  // Make sure that the Serial interface is fast enough, so it doesn't stall the
  // MIDI application
  Serial.begin(1000000);
  Control_Surface.begin();
  Control_Surface.setMIDIInputCallbacks(channelMessageCallback,   //
                                        sysExMessageCallback,     //
                                        sysCommonMessageCallback, //
                                        realTimeMessageCallback); //
}

void loop() {
  Control_Surface.loop();
  if (lcd.getDirty()) {
    Serial << "LCD text updated: " << lcd.getText() << endl;
    lcd.clearDirty();
  }
}
```

**Comment by spamanski** [comment]
Pieter, this is awesome! 

I can see everything I was hoping for: 

![serial](https://github.com/user-attachments/assets/1d133445-6f3c-495a-89c4-72c109d03336)

Now I "only" have to figure out how to send it to the display(s). 
Have to hurry now, but I suspect the next step would be to divide the messages by 7 characters like here: 
https://github.com/tttapa/Control-Surface/issues/76#issuecomment-568935868
And also separating the two lines - I think I saw another example where this was mentioned. 
Then send it to the actual displays. 

It's still very much trial and error for me, basically stitching together code snippets, but this looks very promising. Thanks so much :)

**Comment by spamanski** [comment]
Hi Pieter, I'm sorry, but I'm lost. 

I followed this https://github.com/tttapa/Control-Surface/issues/76#issuecomment-568938037 and edited LCDDisplay.hpp accordingly. 

Then I found the different constructors for LCDDisplay, where this one: 
             

```c++
LCDDisplay(DisplayInterface &display, const MCU::LCD<> &lcd,
               const OutputBank &bank, uint8_t track, uint8_t line,
               PixelLocation loc, uint8_t textSize, uint16_t color) 
```
               
looks like it should give me the possibility to divide the text in two lines (0, 1) and pick which one to send to the display.
What I don't understand is why there are four different versions of this constructor inside LSDDisplay.hpp and how to define which one to use in the sketch. 

At the moment, the sketch works with this: 

```c++
MCU::LCDDisplay lcddisps[] {
  // track (1), position (0, 40), font size (1)
  {display, lcd, bank, 1, {0, 40}, 1, WHITE},
  {display, lcd, bank, 2, {64, 40}, 1, WHITE},
 };
```

If I just add the number for the line in the sketch, like:

```c++
MCU::LCDDisplay lcddisps[] = {
// track (1), line(0), position (0, 40), font size (1)
  {display, lcd, bank, 1, 0, {0, 40}, 1, WHITE},
  {display, lcd, bank, 2, 0, {64, 40}, 1, WHITE}, 
  };
```
  
my arduino will start to blink, not work at all and cause Cubase to crash.  
If I disable the first constructor in LSDDisplay.hpp, the sketch won't compile. 

I also noticed that the things that are displayed at the moment (track names for example) are sent in the second line (line 1) of the MCU display by Cubase. 
So I suspect there must be some definition somewhere to do so, but I haven't found it yet. 

So if you find the time to help me out, I would be very grateful.

**Reply by tttapa** [reply]
Zero is not a valid line number, it should be 1 or 2.

**Reply by spamanski** [reply]
That was it :) 
Thanks so much!
